resources/type-system/inference.md

```diff
@@ -970,6 +970,409 @@ with respect to `L` under constraints `C0`
+  initializing expression of a non-late instance variable, then there is a
+  compile-time error.
+
+- Let `m` be an expression artifact whose runtime behavior is to complete with a
+  value that is the target of the current instance member invocation.
+
+_Expression soundness follows from the fact that within a class, enum, mixin, or
+extension type having interface type `T`, the target of any instance member
+invocation is always an instance of `T`, and within an extension, the target of
+any instance member invocation is always an instance of the extension's "on"
+type._
+
+## Logical boolean expressions
+
+Expression type inference of a logical "and" expression (`e_1 && e_2`) or a
+logical "or" expression (`e_1 || e_2`) produces an expression artifact `m` with
+static type `bool`, where `m` is determined as follows:
+
+- Let `m_1` be the result of performing expression type inference on `e_1`, in
+  context `bool`, and then coercing the result to type `bool`.
+
+- Let `m_2` be the result of performing expression type inference on `e_2`, in
+  context `bool`, and then coercring the result to type `bool`.
+
+- Let `m` be an expression artifact whose runtime behavior is as follows:
+
+  - Execute expression artifact `m_1`, and let `o_1` be the resulting value. _By
+    expression soundness, `o_1` will be an instance of the type `bool`._
+
+  - If the expression is a logical "and" expression and `o_1` is `false`, or the
+    expression is a logical "or" expression and `o_1` is `true`, then `m`
+    completes with the value `o_1`. _Expression soundness follows trivially
+    because the complete value is an instance of `bool`, and the static type is
+    `bool`._
+
+  - Otherwise:
+
+    - Execute the expression artifact `m_2`, and let `o_2` be the resulting
+      value. _By expression soundness, `o_2` will be an instance of the type
+      `bool`._
+
+    - Then, `m` completes with the value `o_2`. _Expression soundness follows
+      trivially because the completed value is an instance of `bool`, and the
+      static type is `bool`._
+
+## Await expressions
+
+Expression type inference of an await expression `await e_1`, in context `K`,
+produces an expression artifact `m` with static type `T`, where `m` and `T` are
+determined as follows:
+
+- Define `K_1` as follows:
+
+  - If `K` is `FutureOr<S>` or `FutureOr<S>?` for some type schema `S`, then let
+    `K_1` be `K`.
+
+  - Otherwise, if `K` is `dynamic`, then let `K_1` be `FutureOr<_>`.
+
+  - Otherwise, let `K_1` be `FutureOr<K>`.
+
+- Let `m_1` be the result of performing expression type inference on `e_1`, in
+  context `K_1`.
+
+- Let `T_1` be the static type of `m_1`.
+
+- Let `T` be `flatten(T_1)`.
+
+- Let `m` be an expression artifact whose runtime behavior is as follows:
+
+  - Execute expression artifact `m_1`, and let `o_1` be the resulting value. _By
+    expression soundness, `o_1` will be an instance of the type `T_1`._
```

Comment:
Technically, "the runtime type of o_1 will be a subtype of `T_1`", since `T_1` may be something like `dynamic`.  (I'm ignoring extension types - to be precise we'd have to account for extension type erasure here).